js/components/MilkdropWindow/index.js

```diff
@@ -0,0 +1,111 @@
+import React from "react";
+import screenfull from "screenfull";
+import butterchurn from "butterchurn";
+import reactionDiffusion2 from "butterchurn-presets/presets/converted/Geiss - Reaction Diffusion 2";
+
+class MilkdropWindow extends React.Component {
+  constructor() {
+    super();
+  }
+  componentDidMount() {
+    const scale = window.devicePixelRatio;
+    const analyserNode = this.props.analyser;
+    this.visualizer = butterchurn.createVisualizer(
+      analyserNode.context,
+      this._canvasNode,
+      {
+        width: this._canvasNode.width * scale,
+        height: this._canvasNode.height * scale
+      }
+    );
+    this.visualizer.connectAudio(analyserNode);
+    this.visualizer.loadPreset(reactionDiffusion2);
+    this._renderViz();
+
+    this._canvasNode.addEventListener("dblclick", () => {
+      this._handleRequestFullsceen();
+    });
+
+    require.ensure(["butterchurn-presets"], require => {
+      const butterchurnPresets = require("butterchurn-presets");
+      const presets = butterchurnPresets.getPresets();
+      const presetKeys = Object.keys(presets);
+      this.cycleInterval = setInterval(() => {
+        const presetIdx = Math.floor(presetKeys.length * Math.random());
+        const preset = presets[presetKeys[presetIdx]];
+        this.visualizer.loadPreset(preset);
+      }, 15000);
+    });
+  }
+  componentWillUnmount() {
+    this._pauseViz();
+    this._stopCycling();
+  }
+  componentDidUpdate() {
+    if (
+      !screenfull.isFullscreen &&
+      (this.props.width !== this.width || this.props.height !== this.height)
+    ) {
+      this.width = this.props.width;
+      this.height = this.props.height;
+      this._setRendererSize(this.width, this.height);
+    }
+  }
+  _renderViz() {
+    this.animationFrameRequest = requestAnimationFrame(() => this._renderViz());
+    this.visualizer.render();
+  }
+  _pauseViz() {
+    if (this.animationFrameRequest) {
+      window.cancelAnimationFrame(this.animationFrameRequest);
+      this.animationFrameRequest = null;
+    }
+  }
+  _stopCycling() {
+    if (this.cycleInterval) {
+      clearInterval(this.cycleInterval);
+      this.cycleInterval = null;
+    }
+  }
+  _setRendererSize(width, height) {
+    const scale = window.devicePixelRatio;
+    this._canvasNode.width = width * scale;
+    this._canvasNode.height = height * scale;
```

Comment:
Optional nit: These could be moved to the render method.

```
<canvas width={this.props.width * scale} ... />
```
I forget if you need to add an explicit `px` to that property.

Reply:
I'll test if 'px' is necessary, can't remember.

Maybe better to set explicitly when we change size, not sure how things are handled if we are constantly updating them to the same value. I guess I can test that too :D

Reply:
Updating to the same value is fine. That's basically React's entire job: Take some new state and figure out what (if anything) actually needs to be updated in the DOM.